File: agents/issue-reconciler/issue_reconciler/api/fetchers/issues.py
from __future__ import annotations

from typing import TypedDict

from issue_reconciler.api.client import GitHubClient
from issue_reconciler.config import REPO_NAME, REPO_OWNER

_QUERY = """
  query OpenIssues($owner: String!, $name: String!, $cursor: String) {
    repository(owner: $owner, name: $name) {
      issues(first: 100, after: $cursor, states: [OPEN]) {
        pageInfo { hasNextPage endCursor }
        nodes {
          id
          number
          title
          labels(first: 20) { nodes { name } }
        }
      }
    }
  }
"""


class OpenIssue(TypedDict):
    id: str  # GraphQL node ID - required by the writer for comment/close mutations
    number: int
    title: str
    labels: list[str]


def fetch_open_issues(client: GitHubClient) -> list[OpenIssue]:
    issues: list[OpenIssue] = []
    cursor = None

    while True:
        data = client.query(_QUERY, {"owner": REPO_OWNER, "name": REPO_NAME, "cursor": cursor})
        page = (data.get("repository") or {}).get("issues")
        if not page:
            break

        for node in page["nodes"]:
            issues.append(
                {
                    "id": node["id"],
                    "number": node["number"],
                    "title": node["title"],
                    "labels": [label["name"] for label in node["labels"]["nodes"]],
                }
            )

        if page["pageInfo"]["hasNextPage"]:
            cursor = page["pageInfo"]["endCursor"]
        else:
            break

    return issues

File: agents/issue-reconciler/api/fetchers/issues.ts
import type { GitHubClient } from '../client.ts';
import { REPO_NAME, REPO_OWNER } from '../../config.ts';

export type OpenIssue = {
  number: number;
  title: string;
  labels: string[];
};

const QUERY = `
  query OpenIssues($owner: String!, $name: String!, $cursor: String) {
    repository(owner: $owner, name: $name) {
      issues(first: 100, after: $cursor, states: [OPEN]) {
        pageInfo { hasNextPage endCursor }
        nodes {
          number
          title
          labels(first: 20) { nodes { name } }
        }
      }
    }
  }
`;

type RawResponse = {
  repository: {
    issues: {
      pageInfo: { hasNextPage: boolean; endCursor: string | null };
      nodes: Array<{
        number: number;
        title: string;
        labels: { nodes: Array<{ name: string }> };
      }>;
    };
  } | null;
};

export async function fetchOpenIssues(client: GitHubClient): Promise<OpenIssue[]> {
  const issues: OpenIssue[] = [];
  let cursor: string | null = null;

  do {
    const data: RawResponse = await client.query<RawResponse>(QUERY, {
      owner: REPO_OWNER,
      name: REPO_NAME,
      cursor,
    });

    const page = data.repository?.issues;
    if (!page) break;

    for (const node of page.nodes) {
      issues.push({
        number: node.number,
        title: node.title,
        labels: node.labels.nodes.map((l) => l.name),
      });
    }

    cursor = page.pageInfo.hasNextPage ? page.pageInfo.endCursor : null;
  } while (cursor);

  return issues;
}
